### PR TITLE
HHH-11495 Reduce Criteria overhead: add implementors cache in MetamodelImpl

### PR DESCRIPTION
**Pull request by caristu**
Proposal to fix https://hibernate.atlassian.net/browse/HHH-11495 in Hibernate 5.3

### CODE REVIEW

**Comment by gsmet** [comment]
Note that you can push your follow-up commits to your branch and they will be included in the PR, no need to create another one :).

File: hibernate-core/src/main/java/org/hibernate/metamodel/internal/MetamodelImpl.java [hunk 1]
```diff
@@ -625,7 +625,7 @@ else if ( result == INVALID_IMPORT ) {
 	 */
 	public String[] getImplementors(String className) throws MappingException {
 
-		if (implementorsCache.containsKey(className)) {
+		if (implementorsCache.get(className) != null) {
```
Comment (gsmet) [inline hibernate-core/src/main/java/org/hibernate/metamodel/internal/MetamodelImpl.java]: Here you're still doing 2 map accesses. Sorry, I should have been more clear: you have to store the first one in a variable then reuse the variable for the test and the return.

File: hibernate-core/src/main/java/org/hibernate/metamodel/internal/MetamodelImpl.java [hunk 2]
```diff
@@ -111,6 +111,8 @@
 
 	private final TypeConfiguration typeConfiguration;
 
+	private final transient Map<String, String[]> implementorsCache = new ConcurrentHashMap<>();
```
Comment (gsmet) [inline hibernate-core/src/main/java/org/hibernate/metamodel/internal/MetamodelImpl.java]: I don't think we need to make it transient, the other similar maps such as `jpaEmbeddableTypeMap` are not. The entity graph one is a bit special.

File: hibernate-core/src/main/java/org/hibernate/metamodel/internal/MetamodelImpl.java [hunk 3]
```diff
@@ -623,12 +625,20 @@ else if ( result == INVALID_IMPORT ) {
 	 */
 	public String[] getImplementors(String className) throws MappingException {
 
+		if (implementorsCache.get(className) != null) {
+			return implementorsCache.get(className);
+		}
+
 		final Class clazz;
+		String[] implementors;
```
Comment (gsmet) [inline hibernate-core/src/main/java/org/hibernate/metamodel/internal/MetamodelImpl.java]: Minor nitpick, but considering the flow, I would only initialize the variable where needed.

**Comment by gsmet** [comment]
@caristu I added a few more comments but all in all, it looks like a good idea.

Could you put together a test case allowing to reproduce the performance issue so that I can test the benefit of this approach in a benchmark?

You just need to include it in our test case template: https://github.com/hibernate/hibernate-test-case-templates/tree/master/orm/hibernate-orm-5 and push it to JIRA.

Thanks!

**Comment by caristu** [comment]
I'm reusing the "implementors" variable to retrieve the cached values.

I can use another variable, if you still prefer ti declare the variable just when needed for the other cases, let me know.

**Comment by gsmet** [comment]
@caristu no it's OK for the variable. Now that I think about it, maybe we should have a simple `getImplementors()` that does a `implementorsCache.computeIfAbsent(...)` and a `doGetImplementors()` that really does the work. This way, we won't compute the map several times if concurrent accesses happen when the value is not yet computed.

This is small tuning, I can also do it while squashing the patch.

Do you think you'll have the time to put together a test case?

**Comment by caristu** [comment]
@gsmet ok, sounds great

yes, I'm working it, I'll try to provide the test asap

**Comment by caristu** [comment]
@gsmet I've just pushed the test case

This is the repo: https://github.com/caristu/hibernate-test-case-templates/tree/criteria-overhead
And this is the test case that can be used to measure time differences with and without the fix: https://github.com/caristu/hibernate-test-case-templates/blob/criteria-overhead/orm/hibernate-orm-5/src/test/java/org/hibernate/bugs/CriteriaOverhead.java

**Comment by gsmet** [comment]
@caristu I squashed your commits and used `computeIfAbsent()` instead.

It indeed makes a very big difference when the model is large. Wondering how we could have ignored it for so long.

Thanks for the patch. I'll ask @yrodiere to take another look before merging.

**Comment by caristu** [comment]
@gsmet thanks for handling the PR on so short notice. Would this patch be included in next 5.3.2 version?It would be great.

**Comment by Sanne** [comment]
Verry nice!

What about memory usage, do we need to cap the cache size and/or is there any opportunity to discard entries?

**Comment by gsmet** [comment]
@Sanne my understanding is that it's naturally capped by the number of entities you have in your application. And even more the ones you are explicitly querying.

So I don't think we will have an issue with this map. Except if a user executes random queries with random entity names but I don't think it makes any sense :).

**Comment by Sanne** [comment]
Thanks @gsmet , sounds reasonable.

_getImplementors()_ is public JPA API tough. We should make a defensive copy of that array :/

**Comment by gsmet** [comment]
@Sanne is it? I don't see it exposed with an Eclipse "Open call hierarchy".

**Comment by yrodiere** [comment]
@gsmet The method implements `org.hibernate.Metamodel#getImplementors`, which is public API. Not JPA, Hibernate-native API. But still public API.

I pushed a few additional commits to address that, and more.

Regarding the patch, it seems to make sense, but I'm not convinced it's really useful in practice. It could even be damageable, for all I know.

The given benchmark has so much variance that I don't see how we could trust it. Here are the results on my machine:

* without the patch:
  ```
  *** Criteria with 1 threads done in 5317 ms ***
  *** Criteria with 2 threads done in 2 ms ***
  *** Criteria with 3 threads done in 1330 ms ***
  *** Criteria with 4 threads done in 952 ms ***
  *** Criteria with 5 threads done in 764 ms ***
  *** Criteria with 6 threads done in 3 ms ***
  *** Criteria with 7 threads done in 636 ms ***
  *** Criteria with 8 threads done in 3 ms ***
  ```
* with patch:
  ```
  *** Criteria with 1 threads done in 3063 ms ***
  *** Criteria with 2 threads done in 1203 ms ***
  *** Criteria with 3 threads done in 770 ms ***
  *** Criteria with 4 threads done in 677 ms ***
  *** Criteria with 5 threads done in 485 ms ***
  *** Criteria with 6 threads done in 395 ms ***
  *** Criteria with 7 threads done in 3 ms ***
  *** Criteria with 8 threads done in 9 ms ***
  ```
* with patch plus an additional commit adding a `clone()` call in `getImplementors(String)` to protect the cache from external changes:
  ```
  *** Criteria with 1 threads done in 4482 ms ***
  *** Criteria with 2 threads done in 1743 ms ***
  *** Criteria with 3 threads done in 1051 ms ***
  *** Criteria with 4 threads done in 861 ms ***
  *** Criteria with 5 threads done in 718 ms ***
  *** Criteria with 6 threads done in 2 ms ***
  *** Criteria with 7 threads done in 441 ms ***
  *** Criteria with 8 threads done in 410 ms ***
  ```

Or is it me? Should I use some specific configuration to make the benchmark more reliable?

**Comment by yrodiere** [comment]
Also, if we really go for this, I think we should only cache actual hits, at least as a first step. We don't know how people will use this API, they could potentially iterate on all the classes in the classpath, and then we'd end up with a bloated cache.

Maybe we should cache only in cases where the class passed as a parameter actually has persisted subtypes?

**Comment by gsmet** [comment]
@yrodiere I force-pushed what I had in the works and applied 2 of your commits on top of that, the third one was already taken care of.

**Comment by gsmet** [comment]
@Sanne I reworked the method with your advises in mind. Could you take a look? Thanks!

**Comment by Sanne** [comment]
Nice safer approach to not catch on CNFE.

Why did you decide to not cache on `implementors.length == 0` ?

**Comment by gsmet** [comment]
@Sanne Yoann's idea. It's to avoid caching elements for classes not resolved as entities e.g. it won't be possible to fill in the cache with random classes.

**Comment by Sanne** [comment]
ah, good point. Let's merge it ?!

**Comment by gsmet** [comment]
Done!

@caristu thanks for having pushed hard for this.
@yrodiere @Sanne thanks for the reviews and ideas, great team effort!

**Comment by Sanne** [comment]
yea, thanks all! A pleasure to see this shaped collaboratively, and big thanks to @caristu for the core idea.

**Comment by caristu** [comment]
Thank you all for the effort!

**Pull request closed without merging**